File: grind75/week2/013-Subsets-Medium/013-Subsets-Medium.cpp
// link: https://leetcode.com/problems/subsets/


class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) 
   {
        vector<vector<int>> ans;
        vector<int> arr;
        backtrack(0, nums, arr, ans);
        return ans;
    }

private:
    void backtrack(int i, vector<int>& nums, vector<int>& arr, vector<vector<int>>& ans) 
{
        if (i == nums.size()) 
        {
            ans.push_back(arr);
            return;
        }

        // Take
        arr.push_back(nums[i]);
        backtrack(i + 1, nums, arr, ans);

        // Don't take (backtrack)
        arr.pop_back();
        backtrack(i + 1, nums, arr, ans);
    }
};

File: grind75/week2/013-Subsets-Medium/013-Subsets-Medium.py
# link: https://leetcode.com/problems/subsets/

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:

        # empty subset is a subset
        ans = []

        def bt(i, arr):
            
            # base case to break our recursion
            if i == len(nums):
                ans.append(arr.copy())
                return
            
            # take
            arr.append(nums[i])
            bt(i + 1, arr)
            
            # don't take (backtrack)
            arr.pop()
            bt(i + 1, arr)

        bt(0, [])
        print(ans)
        return ans

    
        
